### ISSUE TITLE
`prefer-const` barks at looping variables

### ISSUE THREAD

**Issue by thefourtheye** [eslint/eslint#4088]
Let's say if I have code like this

```js
for (let a of [1.3, 2, 3, 4]) {
  console.log(a);
}
```

enabling `prefer-const` rule errors out saying, ```error  `a` is never modified, use `const` instead  prefer-const```. But if we change that `let` to `const`, the value will never change in the loop. Note that this happens only in sloppy and strong mode. In strict mode `const` works fine. (All these are tested with Node.js v4.1.2).

Is it really safe/okay to suggest changing loop variable to `const`?

Related: https://github.com/nodejs/node/pull/3152#discussion_r41547947

ESLint version: 0.24.0

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
This was already tackled: https://github.com/eslint/eslint/pull/2740

**Comment by thefourtheye** [comment]
@gyandeeps If I am not wrong, the tests in the PR only checks if the variable is used in loops, but it doesn't check the values. Let's say eslint suggests `const` and if the value never changes, then it will break the applications.

**Comment by gyandeeps** [comment]
I will let other take on from here as I am not mature enough to tackle ES6 questions yet. :smile:

**Comment by thefourtheye** [comment]
No problem :-) I just wanted to bring this problem to this project's attention. That's all :-)

**Comment by nzakas** [comment]
You're using A very old version of ESLint, be sure to upgrade. :)

Bottom line: you can use const in for-in and for-of loops safely. This is explicitly allowed in the spec and they behave like let. 

For more info on this: https://leanpub.com/understandinges6/read#leanpub-auto-block-binding-in-loops

**Comment by ustccjw** [comment]
@nzakas  now in chrome, const can not be used in for...of.

````
const arr = [1, 2, 3, 4]

for (const i of arr) {
  console.log(i)
}
````

In chrome, will console: `1, 1, 1, 1`,
bebel will console: `1, 2, 3, 4`

**Comment by thefourtheye** [comment]
@ustccjw There are two `const`s. One in strict mode another one in sloppy mode. Use 'use strict;', it will work fine.

**Comment by ustccjw** [comment]
@thefourtheye Thanks a lot !

**Comment by lagden** [comment]
@thefourtheye `let` make more sense! Don't you?
I did some test using Firefox 42.

`const` doesn't work!

with or without `'use strict';`

<img width="329" alt="const" src="https://cloud.githubusercontent.com/assets/130963/11194399/ef7d7b34-8c92-11e5-9cc2-f517e6d6ffbb.png">

<img width="247" alt="let" src="https://cloud.githubusercontent.com/assets/130963/11195136/6ec8cd04-8c97-11e5-969c-a947e64f4209.png">

**Comment by lagden** [comment]
Wow!! never mind!!!
